raft-tech/TANF-app
# Issue 735: [Backend] Implement buildpacks

### Summary of Changes

Resolves #735 

Implements the backend part of  [ADR #11 Implement Cloud.gov Buildpacks](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Architecture%20Decision%20Record/011-buildpacks.md)

+ Updates manifest files to explicitly use a build pack
+ Add new scripts to manage deployments to cloud.gov via `cf` cli utility
+ Adds to common configuration trusted CSRF origins to support future work involving multiple deployments

NOTE: This PR only supplies the scripts and changes needed to be able to use a script to deploy the backend with buildpacks. CI/CD changes to use these scripts will be tracked in https://github.com/raft-tech/tanf-app/issues/844

## Deployment Strategies

There are several new strategies added to our deployment script, similar ones will be implemented for the frontend.

+ rolling
+ bind
+ initial
+ rebuild
+ default 

## rolling

Deploy to a new cloud node, leaving the previous one intact until the new one is spun up. 
This will require enough memory to be allocated for two of the deployed app to be running at once.

## bind

Bind the services the application depends on, update the environment variables
and restage the app.

## initial


There is no app with this name, and the services need to be bound to it for it to work
the app will fail to start once, have the services bind, and then get restaged.

## rebuild

You want to redeploy the instance under the same name
Delete the existing app (with out deleting the services)
and perform the initial deployment strategy.

## default

Tear down the currently running instance of the app, deploy the changed files

### How to Test

1. Login to cloud.gov via cloud foundry CLI
`cf login -a api.fr.cloud.gov  --sso`

2. create a `.env.production` file in `./tdrs-backend` to define the deployments environment. The environment consists of the following variables:

```bash
 ACR_VALUES
 BASE_URL
 CLIENT_ASSERTION_TYPE
 CLIENT_ID
 DJANGO_SECRET_KEY
 FRONTEND_BASE_URL
 OIDC_OP_AUTHORIZATION_ENDPOINT
 OIDC_OP_ISSUER
 OIDC_OP_JWKS_ENDPOINT
 OIDC_OP_LOGOUT_ENDPOINT
 OIDC_OP_TOKEN_ENDPOINT
 OIDC_RP_CLIENT_ID
 PRIVATE_KEY
 JWT_CERT_TEST
 AWS_ACCESS_KEY
 AWS_SECRET_ACCESS_KEY
 AWS_BUCKET
 AWS_REGION_NAME
 AWS_ACCESS_KEY_ID
 BUCKET_NAME
 DJANGO_AWS_ACCESS_KEY_ID
 DJANGO_AWS_SECRET_ACCESS_KEY
 DJANGO_AWS_STORAGE_BUCKET_NAME
 JWT_KEY
 DJANGO_SU_NAME
```

3. Make sure you are in the project root directory
4. Run `bash ./scripts/deploy-backend.sh initial development <app-name>` to an `<app-name>` that does not yet exist.
5. You should see output like the following:
```
Starting deployment for app tdp-backend-dev...
Waiting for app to deploy...

Instances starting...

name:              tdp-backend-dev
requested state:   started
routes:            
last uploaded:     Tue 13 Apr 14:53:11 EDT 2021
stack:             cflinuxfs3
buildpacks:        
	name               version   detect output   buildpack name
	python_buildpack   1.7.34    python          python

type:            web
sidecars:        
instances:       1/1
memory usage:    256M
start command:   ./gunicorn_start.sh
     state     since                  cpu    memory      disk         details
#0   running   2021-04-13T18:53:47Z   0.0%   0 of 256M   474M of 2G   
Mapping route tdp-backend-dev.app.cloud.gov to app tdp-backend-dev in org hhs-acf-prototyping / space tanf-dev as [user]...
OK
```
6. Because this is a new app name it hasn't yet been added to login.gov to allow authentication, but we can still demonstrate the app is running properly by using the `curl` command to check the `auth_check` endpoint as follows
```bash
> curl https://<app-name>.app.cloud.gov/v1/auth_check
 {"authenticated":false}
```
If the app weren't running properly it would return an error
7. Alternatively, you should be able to access the Django admin at the app name you deployed to. For example: `tdp-backend-dev.apps.cloud.gov/admin/`

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] The backend deploy does not deploy a docker container
  - [x] The backend deploy implements Cloud.gov buildpacks

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A infrastructure changes only

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A no UX changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

N/A - automated CI deployment will be completed with https://github.com/raft-tech/tanf-app/issues/844

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
